sycl/include/CL/sycl/ONEAPI/sub_group.hpp

```diff
@@ -437,45 +459,63 @@ struct sub_group {
 #endif
   }
 
+#ifdef __SYCL_DEVICE_ONLY__
+#ifdef __NVPTX__
+  template <int N, typename T, access::address_space Space>
+  sycl::detail::enable_if_t<
+      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value>
+  store(multi_ptr<T, Space> dst, const vec<T, N> &x) const {
+    for (int i = 0; i < N; ++i) {
+      *(dst.get() + i * get_max_local_range()[0] + get_local_id()[0]) = x[i];
+    }
+  }
+#else // __NVPTX__
+  template <int N, typename T, access::address_space Space>
+  sycl::detail::enable_if_t<
+      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
+      N != 1 && N != 3 && N != 16>
+  store(multi_ptr<T, Space> dst, const vec<T, N> &x) const {
+    sycl::detail::sub_group::store(dst, x);
+  }
+
   template <int N, typename T, access::address_space Space>
   sycl::detail::enable_if_t<
       sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
       N == 1>
   store(multi_ptr<T, Space> dst, const vec<T, 1> &x) const {
-#ifdef __SYCL_DEVICE_ONLY__
-#ifdef __NVPTX__
-    dst.get()[get_local_id()[0]] = x[0];
-#else
-    store<T, Space>(dst, x);
-#endif // __NVPTX__
-#else
-    (void)dst;
-    (void)x;
-    throw runtime_error("Sub-groups are not supported on host device.",
-                        PI_INVALID_DEVICE);
-#endif
+    sycl::detail::sub_group::store(dst, x);
   }
 
   template <int N, typename T, access::address_space Space>
   sycl::detail::enable_if_t<
       sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
-      N != 1>
-  store(multi_ptr<T, Space> dst, const vec<T, N> &x) const {
-#ifdef __SYCL_DEVICE_ONLY__
-#ifdef __NVPTX__
-    for (int i = 0; i < N; ++i) {
-      *(dst.get() + i * get_max_local_range()[0] + get_local_id()[0]) = x[i];
-    }
-#else
-    sycl::detail::sub_group::store(dst, x);
+      N == 3>
+  store(multi_ptr<T, Space> dst, const vec<T, 3> &x) const {
+    store<1, T, Space>(dst, x.s0());
+    store<2, T, Space>(dst + get_max_local_range()[0], {x.s1(), x.s2()});
```

Comment:
Related to my other question: Why did you decide to split `store` this way, but not `load`?

Reply:
I did that for optimization. My assumption is that 1 load of 4 elements is faster that 2 load of 1 and 2 elements.
For store it will not work because it will garbage the '4th' element. Will fix load to unify.